pytorch_lightning/trainer/distrib_parts.py

```diff
@@ -434,52 +434,82 @@ def copy_trainer_model_properties(self, model):
+        return self.__transfer_data_to_device(batch, device)
 
-    def transfer_batch_to_gpu(self, batch, gpu_id):
-        return self.__transfer_data_to_device(batch, device='gpu', gpu_id=gpu_id)
+    def transfer_batch_to_gpu(self, batch: Any, gpu_id: int):
+        device = torch.device('cuda', gpu_id)
+        return self.__transfer_data_to_device(batch, device)
 
-    def __transfer_data_to_device(self, batch, device, gpu_id=None):
-        if device == 'tpu' and XLA_AVAILABLE:
-            # base case: object can be directly moved using `to`
-            if callable(getattr(batch, 'to', None)):
-                return batch.to(xm.xla_device())
+    def __transfer_data_to_device(self, batch: Any, device: torch.device):
+        if callable(getattr(batch, 'to', None)):
+            return batch.to(device)
 
-        if device == 'gpu':
-            # base case: object can be directly moved using `cuda` or `to`
-            if callable(getattr(batch, 'cuda', None)):
-                # non_blocking will be ignored if tensor is not pinned.
-                # so we can always set it to True
-                return batch.cuda(gpu_id, non_blocking=True)
+        # when list
+        if isinstance(batch, list):
+            for i, x in enumerate(batch):
+                batch[i] = self.__transfer_data_to_device(x, device)
+            return batch
+
+        # when tuple
+        if isinstance(batch, tuple):
+            # when namedtuple
+            if hasattr(batch, '_fields'):
+                elem_type = type(batch)
+                return elem_type(*(self.__transfer_data_to_device(x, device) for x in batch))
+            else:
+                batch = list(batch)
+                for i, x in enumerate(batch):
+                    batch[i] = self.__transfer_data_to_device(x, device)
+                return tuple(batch)
+
+        # when dict
+        if isinstance(batch, dict):
+            for k, v in batch.items():
+                batch[k] = self.__transfer_data_to_device(v, device)
+
+            return batch
+
+        # check if the model hook can move the data
+        model = self.get_model()
+        if model is not None and self.is_overridden('transfer_batch_to_device', model):
+            batch = model.transfer_batch_to_device(batch, device)
+
+        # nothing matches, return the value as is without transform
+        return batch
+
+    def __transfer_data_to_device(self, batch: Any, device: torch.device):
+
+        if self.is_overriden('transfer_batch_to_device'):
+            return self.get_model().transfer_batch_to_device(batch, device)
 
-            if callable(getattr(batch, 'to', None)):
-                # non_blocking will be ignored if tensor is not pinned.
-                # so we can always set it to True
-                return batch.to(torch.device('cuda', gpu_id), non_blocking=True)
+        # base case: object can be directly moved using `to`
+        if callable(getattr(batch, 'to', None)):
```

Comment:
with `non_blocking` this will probably fail for other devices than gpu due to unsupported kwargs

Reply:
true, I have to rethink that part

Reply:
I check the docs 
https://pytorch.org/docs/1.1.0/tensors.html#torch.Tensor.to
and non-blocking is included in all calls. Do you mean it could break for XLA?

Reply:
I also just ran the [MNIST on TPU](https://colab.research.google.com/drive/1-_LKx4HwAxl5M6xPJmqAAu444LTDQoa3) with this branch installed and it runs fine.